File: src/main/java/core/procedures/math/ToInexact.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigComplex;
import core.scm.BigRational;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

public final class ToInexact extends AFn {

  public ToInexact() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "exact->inexact";
  }

  @Override
  public Number apply1(Object arg) {
    return toInexact(arg);
  }

  public static Number toInexact(Object o) {
    if (o instanceof BigComplex) {
      BigComplex c = ((BigComplex)o);
      return new BigComplex(toInexact(c.getRe()), toInexact(c.getIm()));
    }
    if (o instanceof BigRational) {
      return ((BigRational)o).toBigDecimalInexact();
    }
    if (o instanceof BigInteger) {
      return new BigDecimal(o.toString()).setScale(1, Utils.ROUNDING_MODE);
    }
    if (o instanceof BigDecimal) {
      int scale = Math.max(1, ((BigDecimal)o).scale());
      return ((BigDecimal)o).setScale(scale, Utils.ROUNDING_MODE);
    }
    return ((Number)o).doubleValue();
  }
}

File: src/main/java/core/procedures/math/ToInexact.kt
package core.procedures.math

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Complex
import core.scm.Ratio
import core.utils.Utils
import java.math.BigDecimal
import java.math.BigInteger

class ToInexact : AFn<Number?, Number>(name = "exact->inexact", isPure = true, arity = Exactly(1),
                                       mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?): Number = when (arg) {
        is Complex    -> Complex(invoke(arg.re), invoke(arg.im))
        is Ratio      -> arg.toBigDecimalInexact()
        is BigInteger -> arg.toBigDecimal().setScale(1, Utils.ROUNDING_MODE)
        is BigDecimal -> arg.setScale(maxOf(1, arg.scale()), Utils.ROUNDING_MODE)
        else          -> arg!!.toDouble()
    }
}
